### PR TITLE
libmilter-sendmail: the sendmail mail filter API library

### PR DESCRIPTION
**Pull request by val-kulkov**
Maintainer: me / @val-kulkov
Compile tested: bcm53xx, Buffalo WXR-1900DHP, LEDE Reboot SNAPSHOT r4149-74395d9 (2017-05-18)
Run tested: bcm53xx, Buffalo WXR-1900DHP, LEDE Reboot SNAPSHOT r4149-74395d9 (2017-05-18); the library was successfully used to compile and run the OpenDKIM package; basic tests of the Sendmail executable were successful ("sendmail" package is only built if the relevant sections in Makefile are uncommented)

Description:

The sendmail Mail Filter API (Milter) is designed to allow third-party programs access to mail messages as they are being processed in order to filter meta-information and content.

This library is a prerequisite for the OpenDKIM package.

The proposed Makefile also contains sections for building Sendmail. These sections are commented out because more work is necessary to properly complete porting of Sendmail to LEDE/OpenWrt. The notes in Makefile provide details on what is required to complete the port.
    
It has been verified that Sendmail executables compile and run properly on the target system when the commented sections in Makefile are uncommented.

Signed-off-by: Val Kulkov <val.kulkov@gmail.com>

### CODE REVIEW

File: libs/libmilter/Makefile [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=sendmail
+PKG_VERSION:=8.15.2
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME).$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=http://artfiles.org/sendmail.org/pub/sendmail/ ftp://ftp.cs.berkeley.edu/ucb/sendmail/ http://www.netgull.com/sendmail/ http://sendmail.cybermirror.org/
```
Comment (hauke) [inline libs/libmilter/Makefile]: please use a new file for each URL like this:
PKG_SOURCE_URL:=http://artfiles.org/sendmail.org/pub/sendmail/ \
                      ftp://ftp.cs.berkeley.edu/ucb/sendmail/ \
                      http://www.netgull.com/sendmail/ \
                      http://sendmail.cybermirror.org/

File: libs/libmilter/Makefile [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+PKG_LICENSE:=Sendmail
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+MAKE_PATH:=libmilter
+SM_SITECONFIG:=devtools/Site/site.config.m4
+export MILTER_SOVER=1.0.1
```
Comment (hauke) [inline libs/libmilter/Makefile]: I do not like this export, can you try to add this to the MAKE_FLAGS or something like that

File: libs/libmilter/patches/101-pthreads-stack-size.patch [hunk 3]
```diff
@@ -0,0 +1,34 @@
+Index: sendmail-8.15.2/libmilter/libmilter.h
```
Comment (hauke) [inline libs/libmilter/patches/101-pthreads-stack-size.patch]: Please add a small comment why you need this patch.

File: libs/libmilter/Makefile [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		$(PKG_BUILD_DIR)/libmilter/Makefile.m4
+	echo 'dnl' > $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
+	echo "define(\`confCC', \`"$(TARGET_CC)"')" >> $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
+	echo "define(\`confLDOPTS', \`"$(TARGET_LDFLAGS)"')" >> $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
+	echo "APPENDDEF(\`confDONT_INSTALL_CATMAN')" >> $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
+	echo "APPENDDEF(\`confNO_MAN_BUILD')" >> $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
+	echo "APPENDDEF(\`confINCOWN', \`"$(USER)"')" >> $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
+	echo "APPENDDEF(\`confINCGRP', \`"$(USER)"')" >> $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
+	echo "APPENDDEF(\`confLIBOWN', \`"$(USER)"')" >> $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
+	echo "APPENDDEF(\`confLIBGRP', \`"$(USER)"')" >> $(PKG_BUILD_DIR)/$(SM_SITECONFIG)
```
Comment (hauke) [inline libs/libmilter/Makefile]: This block looks strange why is this needed?

Comment (val-kulkov) [inline libs/libmilter/Makefile]: This block of code changes root ownership of certain destination executables, which is the default config option, to the current user by modifying package configuration so that PKG_INSTALL can be performed under a non-root account. The alternative would be to create a patch to change root ownership to the current user in the package installation script. IMO the first option is more future-proof than the second option. Please advise if I am mistaken about this, and I will introduce a patch to the installation script instead of modifying the config.

File: libs/libmilter/patches/101-pthreads-stack-size.patch (hunk 3 above)
Comment (val-kulkov) [inline libs/libmilter/patches/101-pthreads-stack-size.patch]: Done. Please review [the comment](https://github.com/openwrt/packages/pull/4369/commits/4bce6115cfbde375f50ff84c486445cf3dc07d8b#diff-9c8da989d26bb43cf9df88f9e9848d91).

File: libs/libmilter/Makefile (hunk 1 above)
Comment (val-kulkov) [inline libs/libmilter/Makefile]: Done, please review [the commit](https://github.com/openwrt/packages/pull/4369/commits/4bce6115cfbde375f50ff84c486445cf3dc07d8b#diff-607c0602a87f57834f7ebcd862ee8c5f).

File: libs/libmilter/Makefile (hunk 2 above)
Comment (val-kulkov) [inline libs/libmilter/Makefile]: After considering pros and cons, I decided to hardcode the .so version in sharedlibrary.m4 module. After all, any change in sharedlibrary.m4 file would likely require incrementing the .so version code anyway, therefore it probably makes sense to hardcode the version number in the source file.

**Comment by val-kulkov** [comment]
@hauke : please advise if I should squash two commits into one or leave this submission as it is now.

**Comment by karlp** [comment]
should always be squashed for review comments, no need to preserve things that are undesired.

File: libs/libmilter/Makefile [hunk 5]
```diff
@@ -0,0 +1,80 @@
+#
+# Copyright (C) 2017 Val Kulkov <val.kulkov@gmail.com>
+#
+# This is free software, licensed under the MIT License
+#
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=sendmail
```
Comment (karlp) [inline libs/libmilter/Makefile]: do you really want this?  Isn't it libmilter?

File: libs/libmilter/Makefile (hunk 4 above)
Comment (karlp) [inline libs/libmilter/Makefile]: Given how completely obscure that is, consider at least adding a comment inline explaining the purpose.

File: libs/libmilter/Makefile (hunk 5 above)
Comment (pprindeville) [inline libs/libmilter/Makefile]: Yes and no.  It's part of the sendmail distribution, but it compiles into a standalone library that other MTA's like Postfix can also use.  I would have a sendmail package, but have subpackages like sendmail-libmilter.

File: libs/libmilter/Makefile (hunk 4 above)
Comment (pprindeville) [inline libs/libmilter/Makefile]: No, I'd put this into files/ instead.

Any build-time substitutions could be done with `sed`.

File: libs/libmilter/files/sharedlibrary.m4 [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 1999-2001 Sendmail, Inc. and its suppliers.
```
Comment (pprindeville) [inline libs/libmilter/files/sharedlibrary.m4]: Why is this necessary and why does it have a copyright such as this?

File: libs/libmilter/patches/101-pthreads-stack-size.patch (hunk 3 above)
Comment (pprindeville) [inline libs/libmilter/patches/101-pthreads-stack-size.patch]: But... we're typically not linking against glibc, but instead MUSL or eglibc or uClibc...

**Comment by val-kulkov** [comment]
@pprindeville @champtar @karlp @hauke : the proposed package has been significantly reworked. The major changes are:

1. the package has been renamed from libmilter to libmilter-sendmail, to follow a pattern where the library name goes first, followed by the principal package name, such as libustream-cyassl / libustream-mbedtls / libustream-openssl,
2. Makefile now contains sections for compiling and packaging "sendmail" package. All sections for "sendmail" have been commented out for now because completing the port of Sendmail to LEDE/OpenWrt requires more work. The "sendmail" sections were enabled for testing of the proposed Makefile and patches: the Sendmail executables were compiled successfully and the basic tests of the "sendmail" executable on the target platform were successful.
3. The commented-out "sendmail" sections in the proposed Makefile contain details on what is required to complete porting of Sendmail to LEDE/OpenWrt.
4. I am proposing this submission with commented-out "sendmail" sections to allow building "libmilter-sendmail" package before the work on "sendmail" package is complete. "libmilter-sendmail" is a prerequisite to OpenDKIM. The OpenDKIM PR #4370 is LGTM for @hauke. As soon as this PR is merged, the OpenDKIM PR may be reviewed and merged, too.

Finally, it appears that travis-ci incorrectly marked this latest submission as "build failed". At https://travis-ci.org/openwrt/packages/requests, job 58 for this PR is marked as "Build created successfully" right now, however following the link to job 58 leads to a page where the same job 58 is marked as "build failed". I am not sure how I should fix this problem. Any suggestions are welcome.

**Comment by val-kulkov** [comment]
It turns out that travis-ci checks all packages between the original PR commit and the final (the last) commit and if any package fails, it reports "All checks have failed". Thanks to jow who pointed out to the exact line of code that explains what's going on: https://github.com/openwrt/packages/blob/master/.travis_do.sh#L45

I rebased this PR to the current HEAD and the earlier problem is now gone. However, travis-ci check still fails, now for the following reason:

> WARNING: No feed for package 'sendmail' found, maybe it's already part of the standard packages?
> ...
> make[1]: *** No rule to make target `package/sendmail/download'.  Stop.

I do now understand why there is no rule for package/sendmail/download. After running "rm -rf dl/sendmail.8.15.2.tar.gz", I can successfully build libmilter-sendmail locally by running "make package/feeds/packages/sendmail/{clean,compile} V=s".

**Comment by lynxis** [comment]
off-topic: @val-kulkov the travis check is new. it doesn't look (yet) if a "meta-package" (sendmail) contains other packages. So ignore it for now. I like to fix it, but don't think I can manage to come with something up before next week.
Any idea where we should document the travis check? Where did you look for documentation?

**Comment by val-kulkov** [comment]
@lynxis : thank you for clarification. As I am not yet a frequent contributor here, it did not click in my mind that travis-ci was a new thing here; I had seen it before in other Github repos and it did not occur to me that I was seeing travis-ci here for the first time.

Some initial wrinkles are normal and expected. This present case with "sendmail" being the name of a meta-package while there is no real package called "sendmail" is certainly not a typical case. On the other hand, it is probably a useful test case.

It seems to me that the [Submitting patches](https://lede-project.org/submitting-patches) page is probably the most appropriate place for documenting the travis-ci procedure. If the travis-ci procedure is documented in a separate page, the [Submitting patches](https://lede-project.org/submitting-patches) page should cross-references it from multiple sections on that page, such as "Working with Github", "Squashing commits", "Patch checklist".

**Comment by hnyman** [comment]
> > Any idea where we should document the travis check?

It should be documented in the contributing advice page for this "packages" repo (that this travis instance concerns).
https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md

> It seems to me that the Submitting patches page is probably the most appropriate place for documenting the travis-ci procedure. 

The LEDE "Submitting patches" page might be too far away from the packages. Currently travis has no role for main LEDE source repo, so we would need to explain carefully that which source repo it applies. (only packages, but not Luci, routing, telephony feeds or main LEDE source).

**Comment by val-kulkov** [comment]
To be clear on this, I addressed the requested changes and provided comments as appropriate on July 24, 2017, which is 29 days ago. Nothing is outstanding on my part.

**Comment by champtar** [comment]
@pprindeville is everything ok for you now?

**Comment by lucize** [comment]
can we have this ?

**Comment by jow-** [comment]
Merging this on my responsibility after discussing this in IRC.
